nslsii/temperature_controllers/lakeshore.py

```diff
@@ -0,0 +1,126 @@
+from collections import OrderedDict
+from ophyd import (Component, DynamicDeviceComponent, Device, EpicsSignal,
+                   EpicsSignalRO, PVPositioner)
+
+
+class Lakeshore336(Device):
+    '''An ``ophyd.Device`` class to be used for Lakeshore336 temp controllers.
+
+    This is a class for controlling Lakeshore model 336 temperature controllers
+    The defaults are set to include four sensors (A, B, C and D) and four
+    control loops (1,2,3 and 4) however this is customizable at initialization
+    time via the kwargs ``temperatures`` and ``controls`` respectively. To
+    instantiate the default settings use the command:
+    ``my_controller = Lakeshore336(PVprefix, name='my_controller')``
+
+    Parameters
+    ----------
+    *args: list
+        The arguments that are passed to ``ophyd.Device``.
+    temperatures: list
+        A list of temperature sensor 'channels' to instantiate, the default is
+        ['A','B','C','D'].
+    controls: list
+        A list of heater control 'channels' to instantiate, the default is
+        [1,2,3,4].
+    **kwargs: dict
+        The keyword arguments passed to ``ophyd.Device``.
+    '''
+
+    def _set_fields(fields, device, prefix, **kwargs):
+        '''A function that allows for the components to be dynamically set.
+
+        Parameters
+        ----------
+        fields: list
+            A list of field identifiers to include, for ``_Temperature``
+            components an example is ['A','B','C','D'] and for ``_Control``
+            components an example is [1,2,3,4].
+        device: object
+            An ``ophyd.Device`` or ``ophyd.Signal`` object that shold be used
+            to create the ``ophyd.Component``s specifed in ``fields``.
+        prefix: str
+            The string to prefix to the field identifier in ``fields`` to
+            create the ``ophyd.Component`` ``suffix``.
+        **kwargs: dict
+            The kwargs that are to be passed into the call to ``device``.
+        '''
+        out_dict = OrderedDict()
+        for field in fields:
+            suffix = f'{prefix}{field}'
+            out_dict[f'{field}'] = (device, suffix, kwargs)
+        return out_dict
+
+    class _Temperature(Device):
```

Comment:
I am not sure defining the classes in the class helps with the clairity.  Making them private is good though.

Reply:
I think it is clearer doing it this way, but could be convinced to change it if the consensus is not with me.

Reply:
have made this a factory function which solves this issue